src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageSearch/PackageSearchResultTableRenderer.cs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading.Tasks;
+using NuGet.Configuration;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.CommandLine.XPlat
+{
+    internal class PackageSearchResultTableRenderer : IPackageSearchResultRenderer
+    {
+        private string _searchTerm;
+        private ILoggerWithColor _loggerWithColor;
+        private bool _exactMatch;
+        private const int LineSeparatorLength = 40;
+        private static readonly string SourceSeparator = new('*', LineSeparatorLength);
+
+        public PackageSearchResultTableRenderer(string searchTerm, ILoggerWithColor loggerWithColor, bool exactMatch)
+        {
+            _searchTerm = searchTerm;
+            _loggerWithColor = loggerWithColor;
+            _exactMatch = exactMatch;
+        }
+
+        public async Task Add(PackageSource source, Task<IEnumerable<IPackageSearchMetadata>> completedSearchTask)
+        {
+            IEnumerable<IPackageSearchMetadata> searchResult;
+
+            _loggerWithColor.LogMinimal(SourceSeparator);
+            _loggerWithColor.LogMinimal($"Source: {source.Name} ({source.SourceUri})");
+
+            if (completedSearchTask == null)
+            {
+                _loggerWithColor.LogMinimal(Strings.Error_CannotObtainSearchSource);
+                return;
+            }
+
+            try
+            {
+                searchResult = await completedSearchTask;
+            }
+            catch (FatalProtocolException ex)
+            {
+                // search
+                // Throws FatalProtocolException for JSON parsing errors as fatal metadata issues.
+                // Throws FatalProtocolException for HTTP request issues indicating critical source(v2/v3) problems.
+                _loggerWithColor.LogError(ex.Message);
+                return;
+            }
+            catch (OperationCanceledException ex)
+            {
+                _loggerWithColor.LogError(ex.Message);
+                return;
+            }
+            catch (InvalidOperationException ex)
+            {
+                // Thrown for a local package with an invalid source destination.
+                _loggerWithColor.LogError(ex.Message);
+                return;
+            }
+
+            if (searchResult == null)
+            {
+                _loggerWithColor.LogMinimal(Strings.Error_NoResource);
+                return;
+            }
+
+            var table = new Table(new[] { 0, 2 }, "Package ID", "Latest Version", "Authors", "Downloads");
+
+            if (_exactMatch)
```

Comment:
This is being used to determine whether to use the last result of the task?

I think by completing the task outside of the renderer, you can have cleaner separation.

The renderer doesn't really change the output between exact match and not exact match, it's a matter of the number of elements at hand. 
As such, it shouldn't know about exact match.

Reply:
Fixed